src/mvba/abba/mod.rs

```diff
@@ -1,56 +1,567 @@
+
+                    // send to all parties the message: `(ID, main-vote, r, v, justification, signature share)`
+                    let main_vote_message = Message {
+                        id: self.id.clone(),
+                        action: Action::MainVote(MainVoteAction {
+                            round: self.r,
+                            value,
+                            justification: just,
+                            sig_share,
+                        }),
+                    };
+
+                    self.broadcast(main_vote_message)?;
+                    self.r += 1;
+                    self.decided_value = decided_value;
+                }
+            }
+        }
+
         Ok(())
     }
 
     pub fn is_decided(&self) -> bool {
-        self.state.as_ref().unwrap().context().decided
+        self.decided_value.is_some()
+    }
+
+    fn add_message(&mut self, sender: &NodeId, msg: &Message) -> Result<bool> {
+        match &msg.action {
+            Action::PreVote(action) => {
+                let pre_votes = self.get_mut_pre_votes_by_round(action.round);
+                if pre_votes.contains_key(sender) {
+                    return Ok(false);
+                }
+
+                pre_votes.insert(*sender, action.clone());
+            }
+            Action::MainVote(action) => {
+                let main_votes = self.get_mut_main_votes_by_round(action.round);
+                if main_votes.contains_key(sender) {
+                    return Ok(false);
+                }
+
+                main_votes.insert(*sender, action.clone());
+            }
+        }
+        Ok(true)
+    }
+
+    fn check_message(&mut self, sender: &NodeId, msg: &Message) -> Result<()> {
+        if msg.id != self.id {
+            return Err(Error::InvalidMessage(format!(
+                "invalid ID. expected: {}, got {}",
+                self.id, msg.id
+            )));
+        }
+
+        match &msg.action {
+            Action::PreVote(action) => {
+                // check the validity of the S-signature share on message (ID, pre-vote, r, b)
+                let sign_bytes = self.pre_vote_bytes_to_sign(action.round, &action.value)?;
+                if !self
+                    .pub_key_set
+                    .public_key_share(sender)
+                    .verify(&action.sig_share, &sign_bytes)
+                {
+                    return Err(Error::InvalidMessage("invalid signature share".to_string()));
+                }
+
+                match &action.justification {
+                    PreVoteJustification::RoundOneNoJustification => {
+                        if action.round != 1 {
+                            return Err(Error::InvalidMessage(format!(
+                                "invalid round. expected 1, got {}",
+                                action.round
+                            )));
+                        }
+                    }
```

Comment:
Did you want to check that the pre-vote was for `zero` as well? We only allow no justification for zero votes.